os_migrate/plugins/module_utils/security_group_rule.py

```diff
@@ -4,132 +4,92 @@
-
-    resource = {}
-    params = {}
-    info = {}
-
-    resource[const.RES_PARAMS] = params
-    resource[const.RES_INFO] = info
-    resource[const.RES_TYPE] = const.RES_TYPE_SECURITYGROUPRULE
-
-    params['tags'] = sorted(sdk_sec_rule['tags'])
-
-    # Decide which attrs are param and which are info
-    set_ser_params_same_name(info, sdk_sec_rule, [
+    import const, reference, resource
+
+
+class SecurityGroupRule(resource.Resource):
+
+    resource_type = const.RES_TYPE_SECURITYGROUPRULE
+    sdk_class = openstack.network.v2.security_group_rule.SecurityGroupRule
+
+    info_from_sdk = [
         'id',
         'security_group_id',
         'remote_group_id',
         'project_id',
         'created_at',
         'updated_at',
         'revision_number',
-    ])
-
-    set_ser_params_same_name(params, sec_refs, [
-        'security_group_name',
-        'remote_group_name',
-    ])
-
-    set_ser_params_same_name(params, sdk_sec_rule, [
+    ]
+    params_from_sdk = [
         'description',
         'direction',
         'port_range_max',
         'port_range_min',
         'protocol',
         'remote_ip_prefix',
-    ])
-
-    return resource
-
-
-def security_group_rule_sdk_params(ser_sec_rule, sec_refs):
-    """Create OpenStack SDK parameters dict for creation or update of the
-    OS-Migrate Security group rules `ser_sec_rule`. Use `sec_refs` for
-    name-to-id mappings.
-
-    Returns: Parameters to be fed into `openstack.network.v2.security_group_rule.SecurityGroupRule`
-    """
-    res_type = ser_sec_rule.get(const.RES_TYPE, None)
-    if res_type != const.RES_TYPE_SECURITYGROUPRULE:
-        raise exc.UnexpectedResourceType(const.RES_TYPE_SECURITYGROUPRULE, res_type)
+    ]
 
-    ser_params = ser_sec_rule[const.RES_PARAMS]
-    sdk_params = {}
+    params_from_refs = [
+        'security_group_name',
+        'remote_group_name',
+    ]
 
-    set_sdk_params_same_name(sec_refs, sdk_params, [
-        'remote_group_id',
+    sdk_params_from_refs = [
         'security_group_id',
-    ])
-
-    set_sdk_params_same_name(ser_params, sdk_params, [
-        'description',
-        'direction',
-        'port_range_max',
-        'port_range_min',
-        'protocol',
-        'remote_ip_prefix',
-        'revision_number',
-    ])
-
-    return sdk_params
+    ]
+
+    @classmethod
+    def from_sdk(cls, conn, sdk_resource):
+        obj = super(SecurityGroupRule, cls).from_sdk(conn, sdk_resource)
+        return obj
+
+    def create_or_update(self, conn):
+        refs = self._refs_from_ser(conn)
+        sdk_params = self._to_sdk_params(refs)
+        result = {}
+        result['changed'] = False
+        try:
+            conn.network.create_security_group_rule(**sdk_params)
+            result['changed'] = True
+        except openstack.exceptions.ConflictException:
+            result['msg'] = "This security group rule already exists"
+
+        return result
```

Comment:
This should return a boo whether the resource changed, not a result dict.

I see why you did this but IMO we should not rewrite the contract of the parent class. We should probably drop the 'msg' info here and just return False.